### DISCUSSION TITLE
Error messages encountered

### DISCUSSION THREAD

**Discussion by des2013** [rectorphp/rector#8792] [Q&A] [answered]
While scanning a certain directory, I encountered many errors.  Some seem to interrupt the rule processing, while others do not.
Sample error:
[ERROR] Could not process "olo/emberpay/lib/Cake/Test/Case/Controller/Component/PaginatorComponentTest.php" file, due
         to:
         "Syntax error, unexpected T_NEW, Syntax error, unexpected T_NEW, Syntax error, unexpected T_NEW, Syntax error,
         unexpected T_NEW". On line: 94

Here is my configuration file:
[rector.php.txt](https://github.com/user-attachments/files/16629544/rector.php.txt)

I used the entire list on the first run, then realized that some of the directories were untouched. So, then I commented out all but the one with errors, and the ones before and after it.  The output shows mobi being scanned first, and then olo, even though an error was found in mobi.  But the onboarding directory was not scanned (the one after olo in my list.)  The errors that happen in olo appear to stop the processing of the list of directories.

Complete output with all errors follows.
[rector-output.txt](https://github.com/user-attachments/files/16630073/rector-output.txt)

The line number is always 94 in every error, so I'm assuming 94 is bogus. 
Some of these errors seem to be interrupting the scan of other directories on my list, perhaps halting the scan. 

Is there a way to get rector to ignore warnings and errors, and continue scans regardless?
Can we get the bugs fixed?

**Comment by samsonasik** [comment]
Could you create reproducible at https://getrector.com/demo ? Thank you.

**Reply by des2013** [reply]
demo not feasible as code set is quite large and proprietary,

**Comment by des2013** [comment]
1) No answers.  I discovered by trial and error that rector cannot deal with buggy code.  If there are any syntax errors present in the source code, rector goes nuts and generally dies spewing all manor of indecipherable errors of its own.  The line numbers are useless, but at least it helped me pinpoint which files contained errors, and I debugged those. Once the syntax issues were fixed, rector runs to completion.

2) Examples suggest this question:
Does anyone know what withTypeCoverageLevel does?
reading the rector source code, it has something to do with to what degree the rules are applied. That could not be any more ambiguous.  Inserting that into the mix just adds a new level of chaos.

3) Most of the changes that rector makes to my source code appear to be the insertion of blank lines.  It will pick some apparently random spot, and insert a tab or two and a linefeed into the source file.  No indications in the output what this is or why it is done. Is this an artifact of normal operation, or is this indicating something that could have happened, but is not?

IS ANYONE OUT THERE?  In code space, no one can hear you scream.

**Comment by samsonasik** [comment, marked as answer]
1. rector is not error handler, so that's expected, ensure that your code is valid before run it, using parallel-lint can help.
2. It apply type declaration step by step, use level 0 for the beginning, and move forward
3. we use nikic/php-parser to print code, if you have some mix spaces, that's can happen, see known drawback https://github.com/rectorphp/rector?tab=readme-ov-file#known-drawbacks

I suggest to being slow, step by step, fix the code first, fix spacing (use coding standard tool or right IDE should can handle that), ensure autoload correctly setup, then try with simple rule one rule at a time so that you can easy to check and review, and fix edge cases.

**Reply by des2013** [reply]
thanks.  notes on your enumerated points:
1) bold print notice in the first paragraph would be helpful.  I get it, it's a pattern matcher, and does not parse php. rule-based parser would be better.
2) define: steps, inferring first step is '0'.  what are other steps? 1-100, 1-1000?  What does 'steps' mean in this context? What increments are advised? 1, 10, 25?
3) please explain need for "print code". what is this? why is it needed?  how is it used?  "mix spaces"?
